File: TaskTrackerAPI/Services/TaskPriorityService.cs
/*
 * Copyright (c) 2025 Carlos Abril Jr
 * All rights reserved.
 *
 * This source code is licensed under the Business Source License 1.1
 * found in the LICENSE file in the root directory of this source tree.
 *
 * This file may not be used, copied, modified, or distributed except in
 * accordance with the terms contained in the LICENSE file.
 */
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;
using TaskTrackerAPI.Data;
using TaskTrackerAPI.Models;
using TaskTrackerAPI.Services.Interfaces;
using TaskTrackerAPI.Repositories.Interfaces;
using TaskTrackerAPI.DTOs.Tasks;

namespace TaskTrackerAPI.Services
{
    public class TaskPriorityService : ITaskPriorityService
    {
        private readonly ApplicationDbContext _context;
        private readonly ITaskItemRepository _taskRepository;
        private readonly ITaskService _taskService;

        public TaskPriorityService(ApplicationDbContext context, ITaskItemRepository taskRepository, ITaskService taskService)
        {
            _context = context;
            _taskRepository = taskRepository;
            _taskService = taskService;
        }

        /// <inheritdoc />
        public async Task<TaskItem?> GetHighestPriorityTaskAsync(string userId)
        {
            int userIdInt = int.Parse(userId);
            
            // Get incomplete tasks for the user
            List<TaskItem> tasks = await _context.Tasks
                .Where(t => t.UserId == userIdInt && 
                            t.Status != TaskItemStatus.Completed &&
                            t.Status != TaskItemStatus.Cancelled)
                .ToListAsync();
                
            return tasks
                .OrderByDescending(GetPriorityValue)
                .ThenBy(t => t.DueDate) // Earlier due date first
                .FirstOrDefault();
        }

        /// <inheritdoc />
        public async Task<List<TaskItem>> GetPrioritizedTasksAsync(string userId, int count)
        {
            int userIdInt = int.Parse(userId);
            
            // Get prioritized tasks for the user
            List<TaskItem> tasks = await _context.Tasks
                .Where(t => t.UserId == userIdInt && 
                            t.Status != TaskItemStatus.Completed &&
                            t.Status != TaskItemStatus.Cancelled)
                .ToListAsync();
                
            return tasks
                .OrderByDescending(GetPriorityValue)
                .ThenBy(t => t.DueDate) // Earlier due date first
                .Take(count)
                .ToList();
        }

        /// <summary>
        /// Calculates a priority score for a task based on various factors
        /// </summary>
        private double GetPriorityValue(TaskItem task)
        {
            double priorityValue = 0;

            // Priority factor (High = 3, Medium = 2, Low = 1, default = 0)
            switch (task.Priority?.ToLower())
            {
                case "critical":
                    priorityValue += 4;
                    break;
                case "high":
                    priorityValue += 3;
                    break;
                case "medium":
                    priorityValue += 2;
                    break;
                case "low":
                    priorityValue += 1;
                    break;
                default:
                    priorityValue += 0;
                    break;
            }

            // Due date factor
            if (task.DueDate.HasValue)
            {
                TimeSpan timeUntilDue = task.DueDate.Value - DateTime.UtcNow;
                
                // Overdue tasks get highest priority
                if (timeUntilDue.TotalHours < 0)
                {
                    priorityValue += 5;
                }
                // Due within 24 hours
                else if (timeUntilDue.TotalHours < 24)
                {
                    priorityValue += 4;
                }
                // Due within 3 days
                else if (timeUntilDue.TotalDays < 3)
                {
                    priorityValue += 3;
                }
                // Due within a week
                else if (timeUntilDue.TotalDays < 7)
                {
                    priorityValue += 2;
                }
                // Due within two weeks
                else if (timeUntilDue.TotalDays < 14)
                {
                    priorityValue += 1;
                }
            }

            // Status factor - "In Progress" tasks get a boost
            if (task.Status == TaskItemStatus.InProgress)
            {
                priorityValue += 1;
            }

            return priorityValue;
        }

        /// <summary>
        /// Automatically adjusts the priority of a user's tasks based on due dates, completion patterns, and other factors
        /// </summary>
        /// <param name="userId">The ID of the user whose tasks to adjust</param>
        /// <returns>A summary of priority adjustments made</returns>
        public async Task<PriorityAdjustmentSummaryDTO> AutoAdjustTaskPrioritiesAsync(int userId)
        {
            // Get all incomplete tasks for the user
            IEnumerable<TaskItem> notStartedTasks = await _taskRepository.GetTasksByStatusAsync(userId, TaskItemStatus.NotStarted);
            IEnumerable<TaskItem> inProgressTasks = await _taskRepository.GetTasksByStatusAsync(userId, TaskItemStatus.InProgress);
            
            // Create a new list and add all items from both sources
            List<TaskItem> incompleteTasks = new List<TaskItem>();
            foreach (TaskItem task in notStartedTasks)
            {
                incompleteTasks.Add(task);
            }
            foreach (TaskItem task in inProgressTasks)
            {
                incompleteTasks.Add(task);
            }
            
            PriorityAdjustmentSummaryDTO summary = new PriorityAdjustmentSummaryDTO
            {
                TotalTasksEvaluated = incompleteTasks.Count(),
                TasksAdjusted = 0,
                UpgradedTasks = 0,
                DowngradedTasks = 0,
                AdjustmentTimestamp = DateTime.UtcNow
            };
            
            List<TaskPriorityAdjustmentDTO> adjustments = new List<TaskPriorityAdjustmentDTO>();
            
            foreach (TaskItem task in incompleteTasks)
            {
                TaskPriority calculatedPriority = CalculateIdealPriority(task);
                TaskPriority originalPriority = Enum.Parse<TaskPriority>(task.Priority);
                
                // If the calculated priority differs from the current priority, adjust it
                if (calculatedPriority != originalPriority)
                {
                    TaskItemDTO taskDto = new TaskItemDTO 
                    {
                        Id = task.Id,
                        Title = task.Title,
                        Description = task.Description,
                        Status = task.Status,
                        Priority = (int)calculatedPriority,
                        DueDate = task.DueDate,
                        CategoryId = task.CategoryId,
                        BoardId = task.BoardId,
                        Version = task.Version
                    };
                    
                    // Update the task priority
                    await _taskService.UpdateTaskAsync(userId, task.Id, taskDto);
                    
                    // Record the adjustment
                    adjustments.Add(new TaskPriorityAdjustmentDTO
                    {
                        TaskId = task.Id,
                        TaskTitle = task.Title,
                        PreviousPriority = originalPriority,
                        NewPriority = calculatedPriority,
                        AdjustmentReason = GetAdjustmentReason(task, calculatedPriority)
                    });
                    
                    summary.TasksAdjusted++;
                    
                    if ((int)calculatedPriority > (int)originalPriority)
                        summary.UpgradedTasks++;
                    else
                        summary.DowngradedTasks++;
                }
            }
            
            summary.Adjustments = adjustments;
            return summary;
        }

        /// <summary>
        /// Calculates the ideal priority for a task based on due date, completion patterns, and status
        /// </summary>
        private TaskPriority CalculateIdealPriority(TaskItem task)
        {
            // Default to the current priority if parsing fails
            if (!Enum.TryParse<TaskPriority>(task.Priority, out TaskPriority currentPriority))
            {
                currentPriority = TaskPriority.Medium;
            }
            
            // If no due date, don't change priority
            if (!task.DueDate.HasValue)
                return currentPriority;
                
            double daysUntilDue = (task.DueDate.Value - DateTime.UtcNow).TotalDays;
            
            // Overdue tasks should always be High priority
            if (daysUntilDue < 0)
                return TaskPriority.High;
                
            // Due today or tomorrow -> increase priority
            if (daysUntilDue < 1)
                return TaskPriority.High;
                
            // Due within 3 days -> at least Medium priority
            if (daysUntilDue < 3)
                return currentPriority < TaskPriority.Medium ? TaskPriority.Medium : currentPriority;
                
            // Due in more than 2 weeks and currently High priority -> reduce to Medium
            if (daysUntilDue > 14 && currentPriority == TaskPriority.High)
                return TaskPriority.Medium;
                
            // In all other cases, maintain the current priority
            return currentPriority;
        }

        /// <summary>
        /// Generates a human-readable reason for the priority adjustment
        /// </summary>
        private string GetAdjustmentReason(TaskItem task, TaskPriority newPriority)
        {
            TaskPriority currentPriority = Enum.Parse<TaskPriority>(task.Priority);
            
            if (!task.DueDate.HasValue)
                return "Manual adjustment based on task importance";
                
            double daysUntilDue = (task.DueDate.Value - DateTime.UtcNow).TotalDays;
            
            if (daysUntilDue < 0)
                return "Task is overdue";
                
            if (daysUntilDue < 1)
                return "Task is due today or tomorrow";
                
            if (daysUntilDue < 3 && newPriority > currentPriority)
                return "Task is due within 3 days";
                
            if (daysUntilDue > 14 && newPriority < currentPriority)
                return "Task due date is far in the future";
                
            return "Adjustment based on due date analysis";
        }
    }
} 

File: tasktracker-fe/src/lib/services/taskPriorityService.ts
/*
 * Task Priority Service
 * Handles task priority management, auto-adjustment, and prioritization
 */

import { apiService } from './apiService';
import { ApiResponse } from '@/lib/types/api';
import { 
  TaskPriority,
  PriorityAdjustment,
  PriorityAdjustmentSummary,
  PrioritizedTask
} from '@/lib/types/task';

class TaskPriorityService {
  private baseUrl = '/v1/task-priority';

  /**
   * Automatically adjusts task priorities based on due dates and other factors
   */
  async autoAdjustPriorities(): Promise<ApiResponse<PriorityAdjustmentSummary>> {
    try {
      const response = await apiService.post<PriorityAdjustmentSummary>(
        `${this.baseUrl}/auto-adjust`,
        {}
      );
      return response;
    } catch (error) {
      console.error('Error auto-adjusting priorities:', error);
      throw error;
    }
  }

  /**
   * Gets the highest priority task for the current user
   */
  async getHighestPriorityTask(): Promise<ApiResponse<PrioritizedTask>> {
    try {
      const response = await apiService.get<PrioritizedTask>(
        `${this.baseUrl}/highest-priority`
      );
      return response;
    } catch (error) {
      console.error('Error getting highest priority task:', error);
      throw error;
    }
  }

  /**
   * Gets a list of prioritized tasks for the current user
   */
  async getPrioritizedTasks(count: number = 10): Promise<ApiResponse<PrioritizedTask[]>> {
    try {
      const response = await apiService.get<PrioritizedTask[]>(
        `${this.baseUrl}/prioritized?count=${count}`
      );
      return response;
    } catch (error) {
      console.error('Error getting prioritized tasks:', error);
      throw error;
    }
  }

  /**
   * Manually adjust a task's priority
   */
  async manuallyAdjustTaskPriority(
    taskId: number, 
    newPriority: TaskPriority,
    reason?: string
  ): Promise<ApiResponse<void>> {
    try {
      const response = await apiService.put<void>(
        `/v1/taskitems/${taskId}`,
        {
          priority: newPriority,
          adjustmentReason: reason || 'Manual priority adjustment'
        }
      );
      return response;
    } catch (error) {
      console.error('Error manually adjusting task priority:', error);
      throw error;
    }
  }

  /**
   * Get priority adjustment history for a task
   */
  async getPriorityHistory(taskId: number): Promise<ApiResponse<PriorityAdjustment[]>> {
    try {
      // This would need a new backend endpoint, for now return empty
      return {
        data: [],
        status: 200
      };
    } catch (error) {
      console.error('Error getting priority history:', error);
      throw error;
    }
  }

  /**
   * Calculate priority score for a task (client-side estimation)
   */
  calculatePriorityScore(task: {
    priority: string;
    dueDate?: string;
    status: string;
  }): number {
    let score = 0;

    // Priority factor
    switch (task.priority?.toLowerCase()) {
      case 'critical': score += 4; break;
      case 'high': score += 3; break;
      case 'medium': score += 2; break;
      case 'low': score += 1; break;
      default: score += 0; break;
    }

    // Due date factor
    if (task.dueDate) {
      const dueDate = new Date(task.dueDate);
      const now = new Date();
      const hoursUntilDue = (dueDate.getTime() - now.getTime()) / (1000 * 60 * 60);

      if (hoursUntilDue < 0) {
        score += 5; // Overdue
      } else if (hoursUntilDue < 24) {
        score += 4; // Due today
      } else if (hoursUntilDue < 72) {
        score += 3; // Due within 3 days
      } else if (hoursUntilDue < 168) {
        score += 2; // Due within a week
      } else if (hoursUntilDue < 336) {
        score += 1; // Due within two weeks
      }
    }

    // Status factor
    if (task.status === 'in-progress') {
      score += 1;
    }

    return score;
  }

  /**
   * Get recommended priority for a task
   */
  getRecommendedPriority(task: {
    priority: string;
    dueDate?: string;
    status: string;
  }): { priority: TaskPriority; reason: string } {
    if (!task.dueDate) {
      return {
        priority: this.stringToTaskPriority(task.priority),
        reason: 'No due date set'
      };
    }

    const dueDate = new Date(task.dueDate);
    const now = new Date();
    const daysUntilDue = (dueDate.getTime() - now.getTime()) / (1000 * 60 * 60 * 24);

    if (daysUntilDue < 0) {
      return {
        priority: TaskPriority.High,
        reason: 'Task is overdue'
      };
    } else if (daysUntilDue < 1) {
      return {
        priority: TaskPriority.High,
        reason: 'Task is due today or tomorrow'
      };
    } else if (daysUntilDue < 3) {
      return {
        priority: TaskPriority.Medium,
        reason: 'Task is due within 3 days'
      };
    } else if (daysUntilDue > 14) {
      return {
        priority: TaskPriority.Low,
        reason: 'Task due date is far in the future'
      };
    }

    return {
      priority: this.stringToTaskPriority(task.priority),
      reason: 'Current priority is appropriate'
    };
  }

  /**
   * Convert string priority to TaskPriority enum
   */
  private stringToTaskPriority(priority: string): TaskPriority {
    switch (priority?.toLowerCase()) {
      case 'critical': return TaskPriority.Critical;
      case 'high': return TaskPriority.High;
      case 'medium': return TaskPriority.Medium;
      case 'low': return TaskPriority.Low;
      default: return TaskPriority.Medium;
    }
  }

  /**
   * Convert TaskPriority enum to string
   */
  taskPriorityToString(priority: TaskPriority): string {
    switch (priority) {
      case TaskPriority.Critical: return 'Critical';
      case TaskPriority.High: return 'High';
      case TaskPriority.Medium: return 'Medium';
      case TaskPriority.Low: return 'Low';
      default: return 'Medium';
    }
  }

  /**
   * Get priority color for UI display
   */
  getPriorityColor(priority: string | TaskPriority): string {
    const priorityStr = typeof priority === 'number' 
      ? this.taskPriorityToString(priority)
      : priority;

    switch (priorityStr?.toLowerCase()) {
      case 'critical': return '#dc2626'; // red-600
      case 'high': return '#ea580c'; // orange-600
      case 'medium': return '#ca8a04'; // yellow-600
      case 'low': return '#16a34a'; // green-600
      default: return '#6b7280'; // gray-500
    }
  }

  /**
   * Get priority icon for UI display
   */
  getPriorityIcon(priority: string | TaskPriority): string {
    const priorityStr = typeof priority === 'number' 
      ? this.taskPriorityToString(priority)
      : priority;

    switch (priorityStr?.toLowerCase()) {
      case 'critical': return '🔴';
      case 'high': return '🟠';
      case 'medium': return '🟡';
      case 'low': return '🟢';
      default: return '⚫';
    }
  }
}

export const taskPriorityService = new TaskPriorityService();
export default taskPriorityService; 